### ISSUE TITLE
The fps counters work strangely with DOSBox

### ISSUE THREAD

**Issue by Kappa971** [dosbox-staging/dosbox-staging#1167]
What I've always noticed in DOSBox is that for some reason, the fps are very fluctuating ... For example in any game that runs at 70 fps (or 35 in the case of Doom), the moment you stop playing and on screen there are no moving objects, fps also drop to 2-3fps. I did not report it as a bug because I believe it is a DOSBox feature, but I wanted to ask what is the usefulness of this feature? With a freesync/gsync monitor I don't think it's very useful as they see the fps constantly changing. Thanks for your patience and availability :)

**Comment by kcgen** [comment]
DOSBox has a feature where it emulates fewer CPU cycles when the window looses focus - however the passage of time will still be kept constant so the game will like normal regardless of window-focus state.  

This helps lighten the load and give more of your CPU to the foreground application.  

There is an easy way to disable this if you can build from source (if it's an issue). Simply comment out the call to `CPU_Enable_SkipAutoAdjust();` in src/gui/sdlmain.cpp`.

**Comment by Kappa971** [comment]
`DOSBox has a feature where it emulates fewer CPU cycles when the window looses focus`

Thanks for the reply :) Ok so the cycles go down when the window looses focus and this is a good feature, but this happens even if the game is full screen and you leave the controls for a moment (and in the scene , at that moment there are no moving elements). In this case what are the advantages of this happening?

`There is an easy way to disable this if you can build from source (if it's an issue). Simply comment out the call to in .CPU_Enable_SkipAutoAdjust();src/gui/sdlmain.cpp`

I'll do this test to see what changes, thanks

**Comment by kcgen** [comment]
> but this happens even if the game is full screen and you leave the controls for a moment (and in the scene , at that moment there are no moving elements). 

Thanks for clarifyingz @Kappa971 .
That behavior definitely should not happenz and it's not something I've ever experienced (running on various Linux hardware system and a couple simple Windows desktops). 

This sounds like some kind of aggressive USB power-saving mode kicking in, and in turn causing poll times to go way up.

Is there any way you can disable your USB driver's power save settings for the keyboard and mouse?

**Comment by Kappa971** [comment]
> That behavior definitely should not happen and it's not something I've ever experienced (running on various Linux hardware systems and a couple simple Windows desktops).

To reproduce the problem is very simple: with a Windows PC, run a game with dosbox in full screen like for example Wolfenstein 3D (I use core normal and 35000 cycles), use an fps counter, start the game, while you move it will go normally at 70 fps, then stop, release the controls and you will see the fps drop to 3-5.

> Is there any way you can disable your USB driver's power save settings for the keyboard and mouse?

I tried to disable "USB selective suspend setting" but it doesn't change anything
![Immagine 2021-07-26 101146](https://user-images.githubusercontent.com/62349018/126956284-6605c739-6148-465d-83ad-7a0c83a7f3b5.png)

**Comment by kcgen** [comment]
Thanks for trying those @Kappa971. Whxih versions of Staging does this happen with? And does it also happen with DOSBox SVN and DOSbox-X too?

**Comment by Kappa971** [comment]
> Which versions of Staging does this happen with?

All versions of DOSBox Staging

> And does it also happen with DOSBox SVN and DOSbox-X too?

Yes, even with the latest versions (DOSBox SVN r4459 and DOSBox-X 0.83.15)

**Comment by Burrito78** [comment]
@Kappa971 May i ask how do you display FPS?

**Comment by Kappa971** [comment]
GeForce Experience Overlay, MSI Afterburner or with Fraps (with Fraps it is even more evident)

**Comment by kcgen** [comment]
I wonder if by-passing acceleration using these conf settings makes any difference?

``` ini
[sdl]
windowresolution = medium
output           = texturenb
texture_renderer = software

[render]
aspect   = true
scaler   = none
glshader = none
```

**Comment by Kappa971** [comment]
> I wonder if by-passing acceleration using these conf settings makes any difference?
> 
> ```ini
> [sdl]
> windowresolution = medium
> output           = texturenb
> texture_renderer = software
> 
> [render]
> aspect   = true
> scaler   = none
> glshader = none
> ```

In this way the fps counter does not work (and freesync does not work either as it remains fixed at 144hz)

**Comment by kcgen** [comment]
> In this way the fps counter doesn't work (and freesync doesn't work either as it remains fixed at 144hz)

Correct; the goal is to bypass the accelerated hardware and driver layer, and go through Windows basic graphical rountines.

When you lift your hands off the mouse and keyboard, do framerates plummet like before?

Another way to check - let Quake play through its timedemo without touching the keyboard/mouse. Launch it with:

`quake.exe +timedemo demo1 -nosound -nocdaudio`

Does it run smooth? What FPS does it show at the end of the demo?

**Comment by Kappa971** [comment]
> When you lift your hands off the mouse and keyboard, do framerates plummet like before?

I don't know, as I said in that case the fps counter doesn't work so I don't know if it stays at fixed 70 fps

> Another way to check - let Quake play through its timedemo without touching the keyboard/mouse. Launch it with:

> quake.exe +timedemo demo1 -nosound -nocdaudio

> Does it run smooth? What FPS does it show at the end of the demo?

I tried the Quake demo at 640x480 and show around 77 fps at the end.

**Comment by Burrito78** [comment]
Well, for me it's the other way around (macOS + Logitech G305):

No hands: 100 fps
Wiggling: 54 fps !!

Logitech M100 corded mouse
No hands: 106 fps
Wiggling: 96 fps

BUT the G305 supports 1000 Hz polling rate and if i dial that back to 125 Hz Quake is fine. Otherwise, most people WILL have their polling rate at 1000 if the hardware supports it...

@kcgen Should i open a ticket for that?

**Comment by Kappa971** [comment]
Drops to 54fps can be normal with quake, it's very challenging. What I have described is a little different, try with Wolfenstein 3D. the steam forum link however seems to indicate what is happening

**Comment by kcgen** [comment]
A common pursuit in computing is to avoid doing work that produces the same result, which almost always saves resources: be it CPU time, power consumption, wear-and-tear (in the case of HDDs, media drives, and SSDs cell life), and bandwidth.

Without knowing the internals of the Nvidia driver or the freesync interface, my hunch is that the driver (or GPU) is comparing consecutive frames and only outputing on changes.

This saves bandwidth over the HDMI or DisplayPort cable (which might also be carrying video to other displays or  gigabit Ethernet), and also saves decode work on the display's side: which likely saves power.

The only 'problem' is that the FPS counter appears to be logging only those 'unique' frames per-second (ie: only those changed frames that have been pushed out by the GPU), and is not tallying the duplicate dropped frames. 

If you hide (or ignore) the FPS counter, what you experience with your eyes should be identical - ie: there should be no degradation in what you experience as a user.  

Is that the case?

**Comment by Burrito78** [comment]
> Drops to 54fps can be normal with quake, it's very challenging. What I have described is a little different, try with Wolfenstein 3D. 

Yes, your issue is different. 54fps is normal in Quake, but not in demo1. Thats a pretty controlled situation. 
The massive framerate dips only happen, because the high polling rate of the mouse is bogging down Dosbox somehow. Question is: Is that also normal and expected or does it justify deeper investigations.

**Comment by Kappa971** [comment]
> If you hide (or ignore) the FPS counter, what you experience with your eyes should be identical - ie: there should be no degradation in what you experience as a user.

Exactly, however, in addition to the fps counter, the freesync monitor is also realizing that this is happening, in fact, in its menu the hz fluctuate a lot. When it drops to 2 fps the monitor goes to 50hz, at 70 fps it oscillates at 65-75hz or the lfc is activated, this "jolt" sometimes causes a slight jerk of the video

**Comment by Burrito78** [comment]
> > And btw., Magnet is not running! :)
> 
> You mean the Steam forum link? It should work now, Sorry :)
No, just a reminder of a frustrating problem i solved together with kcgen with the quality Mac app "Magnet".

I think i found my issue, its probably a macOS/Logitech thing.

https://mmazzarolo.com/blog/2021-06-18-macos-windowserver-cpu-spike-mouse/

![Bildschirmfoto 2021-07-26 um 20 40 27](https://user-images.githubusercontent.com/61995128/127042271-84309ba8-7b9b-4e43-98e8-ed4470a65b86.png)

**Comment by Kappa971** [comment]
So, will this behavior be maintained?
I would prefer DOSBox to behave like any other emulator or game. In any case, if there are no plans to make changes, I think I can close the discussion as I think everything has been said. Thanks :)

**Comment by 0mnicydle** [comment]
I would also like to chime in on this and say an option to present every frame regardless of changes would be very welcome to users with variable refresh rate displays. The way things are now, even if it runs smoothly (which it does for the most part in my case), there will be a sort of "flciker" for lack of a better word, or variation in perceived brightness of the display, as the monitor swings wildly between extremes in refresh rate.

**Comment by Kappa971** [comment]
I forgot that some freesync monitors suffered from this problem, which is that as the refresh rate changes, the brightness changes. One more reason for developers to remove this DOSBox behavior (or make it possible to disable it from the configuration file)
